src/sentry/grouping/ingest/grouphash_metadata.py

```diff
@@ -1,24 +1,106 @@
 from __future__ import annotations
 
+import logging
+
+from sentry.eventstore.models import Event
+from sentry.grouping.variants import BaseVariant
 from sentry.models.grouphash import GroupHash
-from sentry.models.grouphashmetadata import GroupHashMetadata
+from sentry.models.grouphashmetadata import GroupHashMetadata, HashBasis
+from sentry.models.project import Project
+
+logger = logging.getLogger(__name__)
+
+GROUPING_METHODS_BY_DESCRIPTION = {
+    # All frames from a stacktrace at the top level of the event, in `exception`, or in
+    # `threads` (top-level stacktraces come, for example, from using `attach_stacktrace`
+    # together with `capture_message`)
+    "stack-trace": HashBasis.STACKTRACE,
+    "exception stack-trace": HashBasis.STACKTRACE,
+    "thread stack-trace": HashBasis.STACKTRACE,
+    # Same as above, but restricted to in-app frames
+    "in-app stack-trace": HashBasis.STACKTRACE,
+    "in-app exception stack-trace": HashBasis.STACKTRACE,
+    "in-app thread stack-trace": HashBasis.STACKTRACE,
+    # The value in `message` or `log_entry`, such as from using `capture_message` or calling
+    # `capture_exception` on a string
+    "message": HashBasis.MESSAGE,
+    # Error type and value, in cases where all frames are ignored by stacktrace rules
+    "exception": HashBasis.MESSAGE,
+    # Error type and value, in cases where there's no stacktrace
+    "in-app exception": HashBasis.MESSAGE,
+    # Fingerprint set by the user, either in the client or using server-side fingerprinting rules
+    "custom fingerprint": HashBasis.FINGERPRINT,
```

Comment:
is it possible at all to delineate between these two? (user-set client vs server fingerprint)

Reply:
That will be included in the `hashing_metadata` - there's a `fingerprint_source` entry.